nixos/modules/installer/tools/nixos-option/nixos-option.cc

```diff
@@ -0,0 +1,599 @@
+using nix::tString;
+using nix::UsageError;
+using nix::Value;
+
+// An ostream wrapper to handle nested indentation
+class Out
+{
+  public:
+    class Separator
+    {};
+    const static Separator sep;
+    enum LinePolicy
+    {
+        ONE_LINE,
+        MULTI_LINE
+    };
+    explicit Out(std::ostream & ostream) : ostream(ostream), policy(ONE_LINE), write_since_sep(true) {}
+    Out(Out & o, std::string const & start, std::string const & end, LinePolicy policy);
+    Out(Out & o, std::string const & start, std::string const & end, int count)
+        : Out(o, start, end, count < 2 ? ONE_LINE : MULTI_LINE)
+    {}
+    Out(Out const &) = delete;
+    Out(Out &&) = default;
+    Out & operator=(Out const &) = delete;
+    Out & operator=(Out &&) = delete;
+    ~Out() { ostream << end; }
+
+  private:
+    std::ostream & ostream;
+    std::string indentation;
+    std::string end;
+    LinePolicy policy;
+    bool write_since_sep;
+    template <typename T> friend Out & operator<<(Out & o, T thing);
+};
+
+template <typename T> Out & operator<<(Out & o, T thing)
+{
+    if (!o.write_since_sep && o.policy == Out::MULTI_LINE) {
+        o.ostream << o.indentation;
+    }
+    o.write_since_sep = true;
+    o.ostream << thing;
+    return o;
+}
+
+template <> Out & operator<<<Out::Separator>(Out & o, Out::Separator /* thing */)
+{
+    o.ostream << (o.policy == Out::ONE_LINE ? " " : "\n");
+    o.write_since_sep = false;
+    return o;
+}
+
+Out::Out(Out & o, std::string const & start, std::string const & end, LinePolicy policy)
+    : ostream(o.ostream), indentation(policy == ONE_LINE ? o.indentation : o.indentation + "  "),
+      end(policy == ONE_LINE ? end : o.indentation + end), policy(policy), write_since_sep(true)
+{
+    o << start;
+    *this << Out::sep;
+}
+
+// Stuff needed for evaluation
+struct Context
+{
+    Context(EvalState * state, Bindings * autoArgs, Value options_root, Value config_root)
+        : state(state), autoArgs(autoArgs), options_root(options_root), config_root(config_root),
+          underscore_type(state->symbols.create("_type"))
+    {}
+    EvalState * state;
+    Bindings * autoArgs;
+    Value options_root;
+    Value config_root;
+    Symbol underscore_type;
+};
+
+Value evaluateValue(Context * ctx, Value * v)
+{
+    ctx->state->forceValue(*v);
+    if (ctx->autoArgs->empty()) {
+        return *v;
+    }
+    Value called{};
+    ctx->state->autoCallFunction(*ctx->autoArgs, *v, called);
+    return called;
+}
+
+bool isOption(Context * ctx, Value const & v)
+{
+    if (v.type != tAttrs) {
+        return false;
+    }
+    auto const & actual_type = v.attrs->find(ctx->underscore_type);
+    if (actual_type == v.attrs->end()) {
+        return false;
+    }
+    try {
+        Value evaluated_type = evaluateValue(ctx, actual_type->value);
+        if (evaluated_type.type != tString) {
+            return false;
+        }
+        return evaluated_type.string.s == static_cast<std::string>("option");
```

Comment:
Shouldn't this be `std::string(evaluated_type.string.s) == "option"`?

Reply:
I don't think it matters.  As long as one or the other of them is a std::string, I get an operator== that looks at content rather than pointer equality.  I picked casting the constant over casting the dynamic thing in hopes that the compiler would have a better chance at optimizing away any runtime cost.

But, I don't think this matters, so I switched it as you suggested.